src/tests/ftest/container/container_async.py

```diff
@@ -21,291 +21,211 @@
   Any reproduction of computer software, computer software documentation, or
   portions thereof marked with this legend must also reproduce the markings.
 '''
-from __future__ import print_function
 import traceback
-import uuid
-import threading
 
-from apricot import TestWithServers, skipForTicket
+from apricot import TestWithServers
 from pydaos.raw import  DaosContainer, DaosApiError
+from test_utils_container import TestContainer
+from test_utils_base import CallbackHandler
 
+RC_SUCCESS = 0
 
-# pylint: disable=global-variable-not-assigned, global-statement
-GLOB_SIGNAL = None
-GLOB_RC = -99000000
 
+class ContainerAsync(TestWithServers):
+    """Tests asynchronous container operations.
 
-def cb_func(event):
-    """
-    Callback function for asynchronous container functionality.
-    """
-    global GLOB_SIGNAL
-    global GLOB_RC
-
-    GLOB_RC = event.event.ev_error
-    GLOB_SIGNAL.set()
-
+    Test create, destroy, open, close, and query.
 
-class ContainerAsync(TestWithServers):
-    """
-    Tests DAOS pool connect permissions (non existing pool handle, bad uuid)
-    and close.
+    Negative case tests the above operations with non-existing DaosContainer.
+    Not all operations are tested for the negative case.
 
     :avocado: recursive
     """
 
     def __init__(self, *args, **kwargs):
         super(ContainerAsync, self).__init__(*args, **kwargs)
-        self.container1 = None
-        self.container2 = None
+        self.container = []
 
-    @skipForTicket("DAOS-4106")
     def test_createasync(self):
-        """
-        Test container create for asynchronous mode.
-
-        :avocado: tags=all,small,full_regression,container,createasync
-        """
+        """Test container create for asynchronous mode.
 
-        global GLOB_SIGNAL
-        global GLOB_RC
+        Test both positive and negative cases. For negative case, RC is -1002,
+        but we just check if it's something other than 0 to make the test
+        robust.
 
-        # initialize a python pool object then create the underlying
-        # daos storage
-        self.prepare_pool()
-        poh = self.pool.pool.handle
+        The negative case is more like a test of the API implementation rather
+        than DAOS itself.
 
-        try:
-            # Container initialization and creation
-            self.container1 = DaosContainer(self.context)
-            self.container2 = DaosContainer(self.context)
+        :avocado: tags=all,small,full_regression,container,cont_create_async
+        """
+        self.add_pool()
+        ph = self.pool.pool.handle
 
-            GLOB_SIGNAL = threading.Event()
-            self.container1.create(poh=poh, con_uuid=None, cb_func=cb_func)
+        cbh1 = CallbackHandler()
+        cbh2 = CallbackHandler()
+        tc1 = TestContainer(pool=self.pool, cb_handler=cbh1)
+        tc2 = TestContainer(pool=self.pool)
+        self.container.append(tc1)
+        self.container.append(tc2)
```

Comment:
Optional: I noticed there is mixed usage of defining `tc1`/`tc2` and then appending them to `self.container` and not using `tc1`/`tc2` in the various `test_*` methods.  It would keep the namespace cleaner to use the later, e.g.
```suggestion
        self.container.append(TestContainer(pool=self.pool, cb_handler=cbh1))
        self.container.append(TestContainer(pool=self.pool))
```

Reply:
Done